pkg/storage/etcd3/compact.go

```diff
@@ -17,13 +17,46 @@ limitations under the License.
 package etcd3
 
 import (
+	"strings"
+	"sync"
 	"time"
 
 	"github.com/coreos/etcd/clientv3"
 	"github.com/golang/glog"
 	"golang.org/x/net/context"
 )
 
+const compactInterval = 10 * time.Minute
+
+var (
+	endpointsMap     map[string]bool
+	endpointsMapLock sync.Mutex
+)
+
+func init() {
+	endpointsMap = make(map[string]bool)
+}
+
+// StartCompactor starts a compactor in the background in order to compact keys
+// older than fixed time.
+// We need to compact keys because we can't let on disk data grow forever.
+// We save the most recent 10 minutes data. It should be enough for slow watchers and to tolerate burst.
+// We might keep a longer history (12h) in the future once storage API can take
+// advantage of multi-version key.
+func StartCompactor(ctx context.Context, client *clientv3.Client, endpoints []string) {
+	endpointsMapLock.Lock()
+	defer endpointsMapLock.Unlock()
+
+	key := strings.Join(endpoints, ";")
```

Comment:
probably just add each endpoint into the map instead of join them

```go
for _, e := range eps {
    if _, ok := endpointsMap[e]; ok {
        glog.V(4)...
        return
    }
}

for _, e := range eps {
    endpointsMap[e] = struct{}
}
```